Make GL fuzz effect emulate streakiness of software

bkoropoff opened the pull request:
The "streak" effect of software fuzz is due to runs of negative offsets in the fuzz offset table causing the same pixel to be copied and darkened several times in a row.  This shader emulates that using a lookup table derived from the original offset table.

Things to make this even more accurate:

- Make the start offset of each column be the sum of non-transparent pixels in all previous columns, rather than random as it is now.
- Actually copy offset background pixels and go through colormaps to perform darkening.  This would require a major change to the rendering pipeline to use intermediate textures so previous pixels can be read.

bkoropoff commented:
![doom01](https://user-images.githubusercontent.com/2101303/236705445-517050e9-8be0-491e-9334-7a961afe6e53.png)

Lovey847 commented:
Here are more accurate multipliers for the fuzz, you'd replace "\<num\>" in the table with "dark\<num\>"
```
const float dark1 = 1.88235294117647058824;
const float dark2 = 3.41176470588235294118;
const float dark3 = 4.62745098039215686275;
const float dark4 = 5.64705882352941176471;
const float dark5 = 6.47058823529411764706;
```

How I calculated them: I used the algorithm `col = (col * (32 - 6) + 16) / 32` (algorithm for calculating the sixth lightmap in dcolors.c, which is the PLAYPAL and COLORMAP generator), and converted from a 0-255 range to what the shader wants.

Lovey847 commented:
There's a slight issue with this algorithm at the moment, if you have 2 identical sprites drawn with fuzz, they'll have the same pattern because the only 2 factors for the RNG are the time passed and the texture X.
![sameFuzz](https://github.com/kraflab/dsda-doom/assets/87508305/7ab06f81-06da-4cad-90f8-d51e1b9eb652)

bkoropoff committed:
Make GL fuzz effect emulate streakiness of software

The "streak" effect of software fuzz is due to runs of negative offsets
in the fuzz offset table causing the same pixel to be copied and
darkened several times in a row.  This shader emulates that using a
lookup table derived from the original offset table, as well as
darkness steps based on dcolors.c

This also changes the "time" uniform to a "count" uniform which is
incremented on each fuzzy sprite drawn, to ensure sprites drawn on
the same frame have different patterns.  The unused screen resolution
uniform was removed.

Things to make this even more accurate:

- Make the start offset of each column be the sum of non-transparent
  pixels in all previous columns, rather than random as it is now.
- Actually copy offset background pixels and go through colormaps
  to perform darkening.  This would require a major change to the
  rendering pipeline to use intermediate textures so previous pixels
  can be read.

bkoropoff commented:
Added the more accurate darkness steps and changed the uniform passed to the shader to a counter to avoid repeating patterns on the same frame.

[Screencast from 2023-05-09 17-32-16.webm](https://github.com/kraflab/dsda-doom/assets/2101303/25492ed9-4b7c-425d-b5e3-1ec694da9ae7)

kraflab commented:
Cool stuff 👍
